### ISSUE TITLE
Request not authorized

### ISSUE THREAD

**Issue by imZack** [Azure/iotedge#13]
## Expected Behavior
Edge module should up and run

## Current Behavior
Edge module will be rejected when calling API to iotedged due to PID mismatch.

Right now, iotedged get pid from docker Pid (e.g. `docker inspect --format '{{.State.Pid}}' moduleName`). https://github.com/Azure/iotedge/blob/master/edgelet/edgelet-docker/src/module.rs#L104 which might not be the process makes the request to daemon.

## Steps to Reproduce
1. Get an edge module
2. Use different process to start connection (e.g. `npm start`)
3. Get rejected

### Device (Host) Operating System
Debian 8

### Architecture
arm32

### Container Operating System
Linux containers

#### iotedged
```
iotedged --version
<6>2018-07-10T05:59:37Z [INFO] - Starting Azure IoT Edge Security Daemon
<6>2018-07-10T05:59:37Z [INFO] - Version - 1.0.0 (52ef77db24126bf473265fc09c53d35290a2dd6b)
iotedged 0.1.0
```

#### Edge Agent
```
iotedge --version
iotedge 1.0.0 (52ef77db24126bf473265fc09c53d35290a2dd6b)
```

#### Docker
```
Client:
 Version:      18.06.0-dev
 API version:  1.37
 Go version:   go1.10.2
 Git commit:   daf021fe
 Built:        Wed Jun 27 00:53:52 2018
 OS/Arch:      linux/arm
 Experimental: false
 Orchestrator: swarm

Server:
 Engine:
  Version:      dev
  API version:  1.37 (minimum version 1.12)
  Go version:   go1.10.2
  Git commit:   9aac187
  Built:        Tue Jun 26 23:27:27 2018
  OS/Arch:      linux/arm
  Experimental: false
```

## Logs
```
Jul 10 12:56:39 Moxa-UC-8112-ME-T-DEMOKIT iotedged[32460]: 2018-07-10T04:56:39Z [INFO] - Request not authorized - expected caller pid: 3599, actual caller pid: 3741
Jul 10 12:56:39 Moxa-UC-8112-ME-T-DEMOKIT iotedged[32460]: 2018-07-10T04:56:39Z [INFO] - Request not authorized - expected caller pid: 3599, actual caller pid: 3741
Jul 10 12:59:32 Moxa-UC-8112-ME-T-DEMOKIT iotedged[32460]: 2018-07-10T04:59:32Z [INFO] - Request not authorized - expected caller pid: 5216, actual caller pid: 5286
Jul 10 12:59:32 Moxa-UC-8112-ME-T-DEMOKIT iotedged[32460]: 2018-07-10T04:59:32Z [INFO] - Request not authorized - expected caller pid: 5216, actual caller pid: 5286
Jul 10 13:00:30 Moxa-UC-8112-ME-T-DEMOKIT iotedged[32460]: 2018-07-10T05:00:30Z [INFO] - Request not authorized - expected caller pid: 5885, actual caller pid: 6054
```

**Comment by myagley** [comment]
Hi YuLun,

This behavior is currently by design. Can you please tell us a little more about your scenario? What is preventing pid 1 in the container from making the calls to `iotedged`?

Thanks!

**Comment by imZack** [comment]
I'm not quite understand why needs to restrict the incoming request by PID of the container instead of checking the pid belongs to the cgroup of the edge module we excepted if we do care where the request should be made.

For the scenario...If you are familiar with application written in node.js, there is a typical bootstrap command called `npm start`. It will forked a process immediately and the pid won't be the same as `docker inspect --format '{{.State.Pid}}' moduleName`. Check this [official docker image of Node](https://github.com/nodejs/docker-node/blob/e3ec2111af089e31321e76641697e154b3b6a6c3/9/onbuild/Dockerfile#L12)

**Comment by myagley** [comment]
Yes, you could make the argument that the design should be changed. :) We are looking at allowing all pids in the particular pid namespace, but we need to think through whether this will preclude anything in the future.

**Comment by ted-poole** [comment]
+1 I have also come across this behaviour while attempting to deploy some legacy/non-azure executables as Modules. I need to have a bit of adaptor script that runs as my module entry point. This adaptor launches an executable (written with the C SDK) to obtain the Module Twin, followed by some processing of the twin's desired properties, and finally it launches the (legacy) executable that implements my business logic.

**Comment by VMinute** [comment]
This is also an issue if you run iotedged from a container (easiest way to run it on the platform I am currently using). It can access docker via socket, but pid of calling process is reported as 0 instead of the expected one.
Having a way to disable this check (leaving it enabled by default) would help.

**Comment by krokoko** [comment]
Any update on this one ? Facing the same issue when following the tutorial to deploy a Azure ML model as an edge module on a device.

**Comment by avranju** [comment]
This change has been merged into the code-base now via PR #858 but it hasn't been included in a release yet though. It should however make it in the next release.

**Comment by myagley** [comment]
This has been release as part of the 1.0.7 release candidate. You can get packages here: https://github.com/Azure/azure-iotedge/releases/tag/1.0.7-rc2
